lib/pbench/server/auth/auth.py

```diff
@@ -138,26 +198,42 @@ def verify_third_party_token(auth_token: str, oidc_client: OpenIDClient) -> bool
         Returns:
             True if the verification succeeds else False
         """
-        identity_provider_pubkey = oidc_client.get_oidc_public_key(auth_token)
         try:
-            oidc_client.token_introspect_offline(
+            identity_provider_pubkey = oidc_client.get_oidc_public_key(auth_token)
+        except Exception:
+            Auth.logger.warning("Identity provider public key fetch failed")
+            identity_provider_pubkey = Auth().get_secret_key()
+        try:
+            audience = oidc_client.client_id if oidc_client else None
+            token_decode = oidc_client.token_introspect_offline(
                 token=auth_token,
                 key=identity_provider_pubkey,
-                audience=oidc_client.client_id,
+                audience=audience,
+                algorithms=algorithms,
                 options={
                     "verify_signature": True,
                     "verify_aud": True,
                     "verify_exp": True,
                 },
             )
-            return True
+            roles = []
+            if audience in token_decode.get("resource_access"):
+                roles = token_decode.get("resource_access").get(audience).get("roles")
```

Comment:
Roles are tied to an *audience* rather than to a *user*?? I clearly don't understand the relationships here.

Reply:
Well, the token belongs to the user so anything encoded in the token belongs to the user. If the user has a `role` under a `client` that will be included in the token. Here, the client is the `audience`. So it doesn't mean that the role is tied to the audience rather a role is tied to a user under the specific audience/client. This is how the example access token payload looks like after decoding:
```
{
  "exp": 1669848367,
  "iat": 1669848067,
  "jti": "d1fcadd9-7aca-45f6-b5d0-bd335467a9f1",
  "iss": "http://localhost:8080/realms/pbench",
  "aud": "account",
  "sub": "6ca9be0b-f7a0-4d5d-a869-bfc9865eb389",
  "typ": "Bearer",
  "azp": "pbench-client",
  "session_state": "fcb219ff-465e-4c66-94d6-5a788ff3f23f",
  "acr": "1",
  "realm_access": {
    "roles": [
      "offline_access",
      "uma_authorization",
      "default-roles-pbench"
    ]
  },
  "resource_access": {
    "pbench-client": {
      "roles": [
        "myrole"
      ]
    },
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
  "scope": "openid email profile",
  "sid": "fcb219ff-465e-4c66-94d6-5a788ff3f23f",
  "email_verified": true,
  "name": "Test Pbench",
  "preferred_username": "test",
  "given_name": "Test",
  "family_name": "Pbench",
  "email": "test@pbench.com"
}
```
Note: The `pbench-client` and `account` are the clients under our `realm`. And under each client, we have roles that this user is tied to. If the user doesn't have any roles under `pbench-client` client then that section will not be included in the `resource_access` key.

Reply:
Cool; but some of that should probably be recorded in the docstring so it's not quite so obscure!